File: src/test/java/com/example/travelator/optional/java/LegsTest.java
package com.example.travelator.optional.java;

import org.junit.jupiter.api.Test;

import java.time.Duration;
import java.util.List;
import java.util.Optional;

import static com.example.travelator.optional.java.Leg.leg;
import static com.example.travelator.optional.java.Legs.findLongestLegOver;
import static java.util.Collections.emptyList;
import static org.assertj.core.api.Assertions.assertThat;

class LegsTest {
	
	private final List<Leg> legs = List.of(
		leg("one hour", Duration.ofHours(1)),
		leg("one day", Duration.ofDays(1)),
		leg("two hour", Duration.ofHours(2))
	);
	
	private final Duration oneDay = Duration.ofDays(1);
	
	@Test
	void is_absent_when_no_legs() {
		Optional<Leg> longestLegOver = findLongestLegOver(emptyList(), Duration.ZERO);
		assertThat(longestLegOver).isEmpty();
	}
	
	@Test
	void is_absent_when_no_legs_enough() {
		assertThat(findLongestLegOver(legs, oneDay)).isEmpty();
	}
	
	@Test
	void is_longest_leg_when_one_match() {
		String actual = findLongestLegOver(legs, oneDay.minusMillis(1))
			.orElseThrow().getDescription();
		assertThat(actual).isEqualTo("one day");
	}
	
	@Test
	void is_longest_leg_when_more_than_one_match() {
		String actual = findLongestLegOver(legs, Duration.ofMinutes(59))
			.orElseThrow().getDescription();
		assertThat(actual).isEqualTo("one day");
	}
}

File: src/test/java/com/example/travelator/optional/kotlin/LegsTest.kt
package com.example.travelator.optional.kotlin

import com.example.travelator.optional.kotlin.Legs.findLongestLegOver
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Disabled
import org.junit.jupiter.api.Test
import java.util.*
import kotlin.time.Duration
import kotlin.time.Duration.Companion.days
import kotlin.time.Duration.Companion.hours
import kotlin.time.Duration.Companion.milliseconds
import kotlin.time.Duration.Companion.minutes

internal class LegsTest {


    private val legs: List<Leg> = listOf(
        Leg.leg("one hour", 1.hours),
        Leg.leg("one day", 1.days),
        Leg.leg("two hour", 2.hours),
    )

    private val oneDay: Duration = 1.days

    @Test
    internal fun testCase1() {
        val actual = findLongestLegOver(emptyList(), Duration.ZERO)
        assertThat(actual).isEmpty
    }

    @Test
    internal fun testCase2() {
        val actual = findLongestLegOver(emptyList(), Duration.ZERO)
        assertThat(actual).isEqualTo(Optional.empty<Any>())
    }

    @Test
    internal fun testCase3() {
        val actual = findLongestLegOver(legs, oneDay)
        assertThat(actual).isEqualTo(Optional.empty<Any>())
    }

    @Test
    internal fun testCase4() { // two hour ??
        val actual = findLongestLegOver(legs, oneDay.minus(1.milliseconds))
            .orElseThrow().description
        assertThat(actual).isEqualTo("one day")
    }

    @Test
    internal fun testCase5() {
        val actual = findLongestLegOver(legs, 59.minutes)
            .orElseThrow().description
        assertThat(actual).isEqualTo("one day")
    }
}